ring/replication_set.go

```diff
@@ -394,6 +395,80 @@ type instanceResult[T any] struct {
 	instance *InstanceDesc
 }
 
+// DoForReplicationSets runs fn for each ring.ReplicationSet in parallel, and returns a list with all merged results.
+// If any invocation returns an error, DoForReplicationSets returns without waiting for the other invocations to finish.
+func DoForReplicationSets[T any](ctx context.Context, sets []ReplicationSet, fn func(context.Context, ReplicationSet) ([]T, error)) ([]T, error) {
+	if len(sets) == 0 {
+		return nil, nil
+	}
+	if len(sets) == 1 {
+		return fn(ctx, sets[0])
+	}
+
+	var (
+		resultsC           = make(chan []T)
+		errorC             = make(chan error)
+		done               = make(chan struct{})
+		fnCtx, cancelFnCtx = context.WithCancel(ctx)
+	)
+
+	defer func() {
+		close(done)
+
+		// Ensure the fnCtx is canceled if it wasn't already.
+		if !errors.Is(fnCtx.Err(), context.Canceled) {
+			cancelFnCtx()
+		}
```

Comment:
cancel funcs they're usually idempotent, not sure we need the check here


```go
// A CancelFunc tells an operation to abandon its work.
// A CancelFunc does not wait for the work to stop.
// A CancelFunc may be called by multiple goroutines simultaneously.
// After the first call, subsequent calls to a CancelFunc do nothing.
type CancelFunc func()
```

Reply:
Closing because outdated.